File: packages/csharp/src/GofPatterns/Behavioral.cs
namespace GofPatterns;

/// <summary>
/// Chain of Responsibility — each handler either answers or calls <c>next()</c>.
/// <code>
/// var route = Chain.Of&lt;Ticket, string&gt;(
///     [(t, next) =&gt; t.Level == 1 ? "bot" : next(),
///      (t, next) =&gt; t.Paid ? "human" : next()],
///     fallback: _ =&gt; "queue");
/// </code>
/// </summary>
public static class Chain
{
    public static Func<TReq, TRes> Of<TReq, TRes>(
        IReadOnlyList<Func<TReq, Func<TRes>, TRes>> handlers,
        Func<TReq, TRes>? fallback = null)
    {
        return request =>
        {
            TRes Step(int i) =>
                i >= handlers.Count
                    ? fallback is not null
                        ? fallback(request)
                        : throw new InvalidOperationException(
                            "Chain: no handler answered and no fallback was given")
                    : handlers[i](request, () => Step(i + 1));

            return Step(0);
        };
    }
}

/// <summary>A command from two delegates, so the simple case needs no class.</summary>
public sealed class Command<T>(Func<T> execute, Action? undo = null)
{
    public Action? Undo { get; } = undo;

    public T Execute() => execute();
}

/// <summary>
/// Command — undoable operations with history. A command with no undo still runs, but
/// cannot be undone.
/// </summary>
public sealed class CommandBus
{
    private readonly Stack<(Func<object?> Execute, Action Undo)> done = new();
    private readonly Stack<(Func<object?> Execute, Action Undo)> undone = new();

    public T Run<T>(Command<T> command)
    {
        var result = command.Execute();
        this.undone.Clear();
        if (command.Undo is { } undo)
        {
            this.done.Push((() => command.Execute(), undo));
        }

        return result;
    }

    public bool Undo()
    {
        if (!this.done.TryPop(out var entry))
        {
            return false;
        }

        entry.Undo();
        this.undone.Push(entry);
        return true;
    }

    public bool Redo()
    {
        if (!this.undone.TryPop(out var entry))
        {
            return false;
        }

        entry.Execute();
        this.done.Push(entry);
        return true;
    }

    public bool CanUndo => this.done.Count > 0;

    public bool CanRedo => this.undone.Count > 0;
}

/// <summary>
/// Observer — one typed channel. <see cref="Subscribe"/> returns an
/// <see cref="IDisposable"/>, so <c>using</c> scopes a subscription.
/// <para><c>event</c> is the language's own answer and is fine; this adds disposable
/// unsubscription and a value you can pass around.</para>
/// </summary>
public sealed class Subject<T>
{
    private readonly List<Action<T>> listeners = [];

    public IDisposable Subscribe(Action<T> listener)
    {
        this.listeners.Add(listener);
        return new Subscription(() => this.listeners.Remove(listener));
    }

    public void Emit(T value)
    {
        // Iterate a copy: a listener may unsubscribe during Emit.
        foreach (var listener in this.listeners.ToArray())
        {
            listener(value);
        }
    }

    public int Count => this.listeners.Count;

    private sealed class Subscription(Action dispose) : IDisposable
    {
        private bool disposed;

        public void Dispose()
        {
            if (!this.disposed)
            {
                this.disposed = true;
                dispose();
            }
        }
    }
}

/// <summary>
/// Mediator — components talk to a hub, never to each other. Channels carry their
/// payload type, so a listener cannot be attached to the wrong event.
/// <code>
/// var invoicePaid = new Mediator.Channel&lt;Invoice&gt;("invoice.paid");
/// hub.On(invoicePaid, invoice =&gt; mailer.Send(invoice.Id));
/// hub.Emit(invoicePaid, invoice);
/// </code>
/// </summary>
public sealed class Mediator
{
    private readonly Dictionary<object, object> subjects = [];

    /// <summary>A named, typed event. The name is only for debugging.</summary>
    public sealed class Channel<T>(string name)
    {
        public string Name { get; } = name;
    }

    private Subject<T> SubjectFor<T>(Channel<T> channel)
    {
        if (!this.subjects.TryGetValue(channel, out var subject))
        {
            subject = new Subject<T>();
            this.subjects[channel] = subject;
        }

        return (Subject<T>)subject;
    }

    public IDisposable On<T>(Channel<T> channel, Action<T> listener) =>
        this.SubjectFor(channel).Subscribe(listener);

    public void Emit<T>(Channel<T> channel, T payload) => this.SubjectFor(channel).Emit(payload);
}

/// <summary>
/// Memento — undo/redo over snapshots of state.
/// <para>Snapshots are stored by reference. Pass <c>snapshot</c> when the state is
/// mutated in place; with a <c>record</c>, a <c>with</c> expression is already a copy.</para>
/// </summary>
public sealed class History<T>(T initial, int limit = int.MaxValue, Func<T, T>? snapshot = null)
{
    private readonly Func<T, T> snapshot = snapshot ?? (value => value);
    private readonly LinkedList<T> past = new();
    private readonly Stack<T> future = new();

    public T Current { get; private set; } = (snapshot ?? (value => value))(initial);

    public void Save(T state)
    {
        this.past.AddLast(this.Current);
        if (this.past.Count > limit)
        {
            this.past.RemoveFirst();
        }

        this.future.Clear();
        this.Current = this.snapshot(state);
    }

    public bool Undo()
    {
        if (this.past.Last is null)
        {
            return false;
        }

        this.future.Push(this.Current);
        this.Current = this.past.Last.Value;
        this.past.RemoveLast();
        return true;
    }

    public bool Redo()
    {
        if (!this.future.TryPop(out var next))
        {
            return false;
        }

        this.past.AddLast(this.Current);
        this.Current = next;
        return true;
    }

    public bool CanUndo => this.past.Count > 0;

    public bool CanRedo => this.future.Count > 0;
}

/// <summary>
/// State — a finite state machine from a transition table.
/// <code>
/// var order = new StateMachine&lt;string, string&gt;("draft", new()
/// {
///     ["draft"] = new() { ["pay"] = "paid" },
///     ["paid"] = new() { ["ship"] = "sent" },
///     ["sent"] = [],
/// });
/// order.Send("pay");  // "paid"
/// </code>
/// <para>States and events are type parameters, so an <c>enum</c> works and a
/// <c>switch</c> over it stays exhaustive.</para>
/// </summary>
public sealed class StateMachine<TState, TEvent>(
    TState initial,
    IReadOnlyDictionary<TState, Dictionary<TEvent, TState>> transitions)
    where TState : notnull
    where TEvent : notnull
{
    public TState State { get; private set; } = initial;

    /// <summary>Emits the state before, the state after, and the event that caused it.</summary>
    public Subject<(TState From, TState To, TEvent Event)> Changes { get; } = new();

    public bool Can(TEvent @event) =>
        transitions.TryGetValue(this.State, out var allowed) && allowed.ContainsKey(@event);

    /// <summary>Returns the new state. Throws on an event the current state disallows.</summary>
    public TState Send(TEvent @event)
    {
        if (!transitions.TryGetValue(this.State, out var allowed)
            || !allowed.TryGetValue(@event, out var target))
        {
            throw new InvalidOperationException(
                $"StateMachine: \"{@event}\" is not allowed in \"{this.State}\"");
        }

        var from = this.State;
        this.State = target;
        this.Changes.Emit((from, target, @event));
        return target;
    }
}

/// <summary>
/// Visitor — dispatch on a node's tag instead of a switch in every function.
/// <para>
/// Pattern matching on a type hierarchy is better when your nodes are classes, and the
/// compiler checks it — prefer that. This dispatches on a value you extract, which is
/// what you need for a <c>JsonNode</c> or a dictionary that never became a class.
/// </para>
/// </summary>
public static class Visitor
{
    public static Func<TNode, TResult> On<TNode, TTag, TResult>(
        Func<TNode, TTag> tag,
        IReadOnlyDictionary<TTag, Func<TNode, TResult>> visitors,
        Func<TNode, TResult>? fallback = null)
        where TTag : notnull
    {
        return node =>
        {
            var key = tag(node);
            if (visitors.TryGetValue(key, out var visit))
            {
                return visit(node);
            }

            return fallback is not null
                ? fallback(node)
                : throw new KeyNotFoundException($"Visitor: no visitor for \"{key}\"");
        };
    }
}

/// <summary>
/// Iterator — turn an external cursor into something <c>foreach</c> and LINQ can walk.
/// <para>
/// Drivers and SDKs hand you a <c>HasNext()</c> / <c>Next()</c> pair. This is lazy, so
/// a <c>break</c> or a <c>First</c> stops pulling. Writing the source yourself? Use
/// <c>yield return</c> — no helper needed.
/// </para>
/// </summary>
public static class Iterator
{
    public static IEnumerable<T> Iterate<T>(Func<bool> hasNext, Func<T> next)
    {
        while (hasNext())
        {
            yield return next();
        }
    }
}

/// <summary>
/// Template Method — a fixed skeleton with replaceable steps.
/// <para>
/// Optional parameters with delegate defaults are the idiomatic version and you should
/// prefer them: <c>Report(Func&lt;string&gt;? read = null)</c>. This exists so the
/// catalog reads the same across languages.
/// </para>
/// </summary>
public static class Template
{
    public static Func<TArg, TResult> Of<THooks, TArg, TResult>(
        THooks defaults,
        Func<THooks, TArg, TResult> skeleton,
        THooks? overrides = null)
        where THooks : class =>
        arg => skeleton(overrides ?? defaults, arg);
}

File: packages/kotlin/src/main/kotlin/io/github/andrewtellez/gof/Behavioral.kt
package io.github.andrewtellez.gof

/**
 * Chain of Responsibility — each handler either answers or calls `next()`.
 *
 * ```
 * val route = chain<Ticket, String>(
 *     listOf(
 *         { t, next -> if (t.level == 1) "bot" else next() },
 *         { t, next -> if (t.paid) "human" else next() },
 *     ),
 *     fallback = { "queue" },
 * )
 * ```
 */
public fun <Req, Res> chain(
    handlers: List<(Req, () -> Res) -> Res>,
    fallback: ((Req) -> Res)? = null,
): (Req) -> Res = { request ->
    fun step(i: Int): Res =
        if (i >= handlers.size) {
            fallback?.invoke(request)
                ?: error("chain: no handler answered and no fallback was given")
        } else {
            handlers[i](request) { step(i + 1) }
        }
    step(0)
}

/**
 * A command built from two lambdas, so the simple case needs no class of its own.
 * A command with no [undo] still runs, but cannot be undone.
 */
public class Command<out T>(
    private val execute: () -> T,
    public val undo: (() -> Unit)? = null,
) {
    public operator fun invoke(): T = execute()
}

/**
 * Command — undoable operations with history.
 *
 * ```
 * val bus = CommandBus()
 * bus.run(Command({ cart.add(sku) }, undo = { cart.remove(sku) }))
 * bus.undo()
 * bus.redo()
 * ```
 */
public class CommandBus {
    private val done: ArrayDeque<Command<*>> = ArrayDeque()
    private val undone: ArrayDeque<Command<*>> = ArrayDeque()

    public fun <T> run(command: Command<T>): T {
        val result = command()
        undone.clear()
        if (command.undo != null) done.addLast(command)
        return result
    }

    public fun undo(): Boolean {
        val command = done.removeLastOrNull() ?: return false
        command.undo?.invoke()
        undone.addLast(command)
        return true
    }

    public fun redo(): Boolean {
        val command = undone.removeLastOrNull() ?: return false
        command()
        done.addLast(command)
        return true
    }

    public val canUndo: Boolean get() = done.isNotEmpty()
    public val canRedo: Boolean get() = undone.isNotEmpty()
}

/**
 * Observer — one typed channel. [subscribe] returns the unsubscribe function.
 *
 * ```
 * val priceChanged = Subject<Double>()
 * val off = priceChanged.subscribe(::render)
 * priceChanged.emit(9.99)
 * off()
 * ```
 */
public class Subject<T> {
    private val listeners: MutableList<(T) -> Unit> = mutableListOf()

    public fun subscribe(listener: (T) -> Unit): () -> Unit {
        listeners += listener
        return { listeners.remove(listener) }
    }

    public fun emit(value: T) {
        // Iterate a copy: a listener may unsubscribe during emit.
        for (listener in listeners.toList()) listener(value)
    }

    public val size: Int get() = listeners.size
}

/**
 * Mediator — components talk to a hub, never to each other.
 *
 * Channels carry their payload type, so a listener cannot be attached to the wrong
 * event and `emit` cannot be given the wrong payload:
 *
 * ```
 * val invoicePaid = Mediator.Channel<Invoice>("invoice.paid")
 * hub.on(invoicePaid) { invoice -> mailer.send(invoice.id) }
 * hub.emit(invoicePaid, invoice)
 * ```
 */
public class Mediator {
    /** A named, typed event. Declare it once and share it; the name is only for debugging. */
    public class Channel<T>(public val name: String)

    private val subjects: MutableMap<Channel<*>, Subject<*>> = mutableMapOf()

    @Suppress("UNCHECKED_CAST")
    private fun <T> subjectFor(channel: Channel<T>): Subject<T> =
        subjects.getOrPut(channel) { Subject<T>() } as Subject<T>

    public fun <T> on(channel: Channel<T>, listener: (T) -> Unit): () -> Unit =
        subjectFor(channel).subscribe(listener)

    public fun <T> emit(channel: Channel<T>, payload: T): Unit = subjectFor(channel).emit(payload)
}

/**
 * Memento — undo/redo over snapshots of state.
 *
 * ```
 * val doc = History(text, limit = 50)
 * doc.save(next)
 * doc.undo()
 * ```
 *
 * Snapshots are stored by reference. Pass [snapshot] — a copy function — when the
 * state is mutated in place; with a `data class`, `copy()` is already what you want.
 */
public class History<T>(
    initial: T,
    private val limit: Int = Int.MAX_VALUE,
    private val snapshot: (T) -> T = { it },
) {
    private val past: ArrayDeque<T> = ArrayDeque()
    private val future: ArrayDeque<T> = ArrayDeque()

    public var current: T = snapshot(initial)
        private set

    public fun save(state: T) {
        past.addLast(current)
        if (past.size > limit) past.removeFirst()
        future.clear()
        current = snapshot(state)
    }

    public fun undo(): T? {
        val previous = past.removeLastOrNull() ?: return null
        future.addLast(current)
        current = previous
        return current
    }

    public fun redo(): T? {
        val next = future.removeLastOrNull() ?: return null
        past.addLast(current)
        current = next
        return current
    }

    public val canUndo: Boolean get() = past.isNotEmpty()
    public val canRedo: Boolean get() = future.isNotEmpty()
}

/**
 * State — a finite state machine from a transition table.
 *
 * ```
 * val order = StateMachine(
 *     initial = "draft",
 *     transitions = mapOf(
 *         "draft" to mapOf("pay" to "paid"),
 *         "paid" to mapOf("ship" to "sent"),
 *         "sent" to emptyMap(),
 *     ),
 * )
 * order.send("pay")  // "paid"
 * order.can("ship")  // true
 * ```
 *
 * States and events are type parameters, so a `sealed interface` or an `enum` works
 * and the compiler checks exhaustiveness where you `when` over them.
 */
public class StateMachine<S : Any, E : Any>(
    initial: S,
    private val transitions: Map<S, Map<E, S>>,
) {
    public var state: S = initial
        private set

    /** Emits every transition: the state before, the state after, and the event. */
    public val changes: Subject<Change<S, E>> = Subject()

    public data class Change<S, E>(val from: S, val to: S, val event: E)

    public fun can(event: E): Boolean = transitions[state]?.containsKey(event) == true

    /** Returns the new state. Throws on an event the current state does not allow. */
    public fun send(event: E): S {
        val target = transitions[state]?.get(event)
            ?: throw IllegalStateException("StateMachine: \"$event\" is not allowed in \"$state\"")
        val from = state
        state = target
        changes.emit(Change(from, state, event))
        return state
    }
}
